### PR TITLE
Make lint diagnostics responsible for providing their primary span

### LINKED ISSUE #125169: `rustc_macros`: Make it possible to derive both `Diagnostic` and `LintDiagnostic` on the same type
In #117164 I had to copy the diagnostic structs `TyParamFirstLocal` and `TyParamSome` and define the separate structs `TyParamFirstLocalLint` and `TyParamSomeLint` to be able to use the translatable diagnostics `hir_analysis_ty_param_first_local` and `hir_analysis_ty_param_some` as both a `Diagnostic` and a `LintDiagnostic`:

https://github.com/rust-lang/rust/blob/b21b74b5e673b9102502e8f69f378313d901746c/compiler/rustc_hir_analysis/src/errors.rs#L1360-L1406

In #116829, had to copy the diagnostic struct `ReprConflicting` and define the separate struct `ReprConflicting` to be able to use the translatable diagnostic `passes_repr_conflicting` as both a `Diagnostic` and a `LintDiagnostic`:

https://github.com/fmease/rust/blob/d0b99e3efe62b5acc107da9c84331eae6cbe5a0d/compiler/rustc_passes/src/errors.rs#L561-L570

In all three cases, I could've probably turned the derivation of `LintDiagnostic` into a manual impl but that doesn't seem very enticing either for various reasons.

For context, this situation arises whenever we want to emit a diagnostic for something but can't report a hard error unconditionally in all cases for backward compatibility and therefore have to emit the very same diagnostic at different “levels of severity” depending on a set of condition. With *level of severity* I'm specifically referring to the set {hard error, lint} here (where *lint* has its own level of course).

---

More concretely, the reason why you can't just `#[derive(Diagnostic, LintDiagnostic)] struct Ty/*...*/` is because of `#[primary_span]`: If the diagnostic struct contains a `#[primary_span]` (which it does most of the time) for the derive macro `Diagnostic`, the derive macro `LintDiagnostic` rejects `#[primary_span]` rendering the two derives incompatible with each other.

Individually, `LintDiagnostic` rejecting `#[primary_span]` makes sense because the span(s) for a lint are to be provided to the function that emits the lint like `TyCtxt::emit_spanned_lint`.

---

There are probably multiple ways to approach this but I'm not super familiar with internals of rustc's linting API. Anyways, I'd like to see this “just work” because this won't be the last time this case will arise.

### PR DESCRIPTION
**Pull request by fmease**
### Summary and Rustc-Dev-Facing Changes

Instead of passing the primary span of lint diagnostics separately — namely to the functions `tcx.emit_span_lint` and `tcx.emit_node_span_lint` —, make lint diagnostics responsible for “storing” (providing) their span. I've removed the two aforementioned methods. You are now expected to use `tcx.emit_lint` and `tcx.emit_node_lint` respectively and to provide the primary span when deriving `LintDiagnostic` (via `#[primary_span]`) / implementing `LintDiagnostic` manually (via `LintDiagnostic::span`).

> [!NOTE]
> FIXME(fmease): Mention how early lints are handled.

### Motivation

Making the general format of `#[derive(LintDiagnostic)]` identical to `#[derive(Diagnostic)]`. I bet this has lead to slight confusion or annoyances before. Moreover, it enables deriving both traits at once (see #125169 for the motivation).

### Approach

As outlined in https://github.com/rust-lang/rust/issues/125169#issuecomment-2116158144, the naïve approach of doing the same as `Diagnostic` doesn't work for `LintDiagnostic`, i.e., setting the primary span with `Diag::span` inside `decorate_lint` (that's `into_diag` for `Diagnostic`) because `lint_level` (`lint_level_impl`) needs access to the primary spans before decorating the `Diag` to be able to filter out some lints (namely if they come from an external macro) etc.

Therefore, I had to introduce a new method to `LintDiagnostic`: `fn span(&self) -> Option<MultiSpan>` (similar to the existing method `LintDiagnostic::msg`). 

### Commits

1. The 1st commit addresses [#125169 (comment)](https://github.com/rust-lang/rust/issues/125169#issuecomment-2113631272). It contains the necessary changes to the linting APIs. It doesn't build on its own due to the removed APIs. Split out for easier reviewing.
2. The 2nd commit updates all existing lint diagnostics to use the new APIs. Most of them are mindless, monotonous, obvious changes. Some changes are *slightly* more involved out of necessity. I've verified (partly programmatically, partly manually) that all lint diags that used to have a primary span still have a primary span.
3. The 3rd commit updates the fulldeps UI tests that exercise the (lint) diagnostic API  
4. The 4th commit fixes [#125169](https://github.com/rust-lang/rust/issues/125169). I've only deduplicated the lint diagnostic structs mentioned in the issue and haven't gone looking for other structs that may benefit from deriving both `Diagnostic` and `LintDiagnostic` because I didn't have the time to do so yet.

### Meta

Fixes #125169.

I'll submit a rustc-dev-guide PR later.
I hope this doesn't need an MCP, it's pretty straight forward.

cc @davidtwco
r? @nnethercote or anyone on compiler who has the energy to review 82 files

### CODE REVIEW

File: compiler/rustc_errors/src/diagnostic.rs
Comment (fmease) [inline compiler/rustc_errors/src/diagnostic.rs:201]: `decorate_lint` is only responsible for decorating a lint diagnostic. It's not responsible for actually emitting the diagnostic!

File: compiler/rustc_errors/src/diagnostic.rs
Comment (fmease) [inline compiler/rustc_errors/src/diagnostic.rs:205]: Open to rename this to `primary_span`.

File: compiler/rustc_macros/src/diagnostics/diagnostic.rs
Comment (fmease) [inline compiler/rustc_macros/src/diagnostics/diagnostic.rs:159]: Initially I tried to implement this in a DRY way reusing existing functions from `diagnostic_builder` but in the end I opted for just doing it the straightforward way because it wasn't entirely obvious to me how I should refactor `diagnostic_builder` to suit my needs.

**Comment by Xiretza** [comment]
I foresee lots of merge conflicts with #124417 :/

**Comment by fmease** [comment]
@Xiretza Happy to block this PR on yours and do the rebasing myself here.

**Comment by nnethercote** [comment]
> @Xiretza Happy to block this PR on yours and do the rebasing myself here.

@fmease: this seems fine to me but I guess you're still waiting on this.

**Review by davidtwco — APPROVED** [review]
This looks good to me, r=me once you're ready to see this merged

**Comment by Dylan-DPC** [comment]
#124417 is merged so unblocking this

**Comment by alex-semenyuk** [comment]
> #124417 is merged so unblocking this

^^ @fmease

**Comment by nnethercote** [comment]
David gave a soft approval for an earlier version of this, so it makes sense to officially hand it over now that it seems close to being ready.

r? @davidtwco

**Comment by fmease** [comment]
Yesterday I coincidentally stumbled upon https://rust-lang.zulipchat.com/#narrow/channel/336883-i18n/topic/.23100717.20diag.20translation/near/472701978 (2024-09-25) which essentially declares the current diagnostic translation effort (#100717) unfruitful and unmaintained. While only I read this post yesterday, I have been aware of this fact for some time.

Therefore I don't want to burden davidtwco with re-reviewing >1000 lines and I'm also fine with closing this PR.

Of course, not a lot has changed. It's essentially a rebase from a 5 month old master. I stalled the rebase until now because I knew that this approach (of making lint diagnostics responsible or providing their primary span) does not scale to early (buffered) lints. I came to realize that when reviewing PR #124417 (which made early lints translatable) and implementing PR #125410 (which won back some of the perf regressions caused by the former).

You say it doesn't scale to early (buffered) lints, so how did you succeed in rebasing then while keeping perf in check? Well, I cheated a little bit. The statement "Lints diagnostics are responsible for providing their primary span" now only only holds for non-early lints. The early lint diagnostics don't provide a `#[primary_span]` / return `None` for `LintDiagnostic::span(&self) -> Option<MultiSpan>` with the **hidden assumption/knowledge**[^1] that it's supplied separately in `opt_span_lint_with_diagnostics` from `BufferedEarlyLint.span`.

For perf reasons, we can't eagerly "lower" `BuiltinLintDiag` to lint diagnostics ("`impl LintDiagnostic`") inside `opt_span_lint_with_diagnostics` via `decorate_lint`[^2]. Only if we lower eagerly, we can truly make early lint diagnostics responsible for providing their primary spans. However, that's off the table. Re. perf, the lowering step is expensive and therefore we delay this step as much as possible. Namely we don't want to lower lint diagnostics that end up not getting emitted (due to lint levels). We achieve this by letting `lint_level` decide which takes a `Span` (which is used in its decision process)! We need the primary span(s) before lowering the lint diagnostics, therefore lint diagnostics cannot possibly provide their primary span!

This creates a divide between early and late `LintDiagnostic`s which is unacceptable in my opinion. I wanted bring `{,Lint}Diagnostic` closer together for ease of understanding and use. This simply doesn't work if `LintDiagnostic` itself is divided. As such, I no longer believe in this PR.

[^1]: Well, I did add comments in the `LintDiagnostic::span` cases but not in the `#[derive(LintDiagnostic)]` ones (there are simply too many).
[^2]: See the merged commit https://github.com/rust-lang/rust/pull/125410/commits/37bf2d2dabdbdce9473b0fed1708fcbf31ba9c1a which reverts this.

**Pull request closed without merging**
